core/drivers/fabric-driver/server/listener.ts

```diff
@@ -5,165 +5,179 @@
                 }
             })
-        })
-    };
-    await network.addBlockListener(listener);
-    channelBlockListenerMap.set(channelId, listener);
-    logger.info(`Added block listener for channel ${channelId}`);
-    return listener;
+        }
+    })
 }
 
 /**
- * Register a chaincode event listener with a callback
+ * Register a block listener with a callback
  **/
-const initContractEventListener = (
-    contract: Contract,
+const initBlockEventListenerForChannel = async (
+    network: Network,
     networkName: string,
     channelId: string,
-    chaincodeId: string,
-): any => {
-    const listener: ContractListener = async (event: ContractEvent) => {
-        logger.info(`Event Listener: Trying to find match for channel: ${channelId}, chaincode: ${chaincodeId}, event class: ${event.eventName}`);
-        // Find all matching event subscriptions stored in the database
-        let eventMatcher = new events_pb.EventMatcher();
-        eventMatcher.setEventType(events_pb.EventType.LEDGER_STATE);
-        eventMatcher.setEventClassId(event.eventName);
-        eventMatcher.setTransactionLedgerId(channelId);
-        eventMatcher.setTransactionContractId(chaincodeId);
-        eventMatcher.setTransactionFunc('*');
-        let eventSubscriptionQueries;
-        for (let i = 0 ; i < 10 ; i++) {
-            try{
-                eventSubscriptionQueries = await lookupEventSubscriptions(eventMatcher);
-                break;
-            } catch(error) {
-                let errorString: string = error.toString();
-                if (!(error instanceof DBLockedError)) {    // Check if contention was preventing DB access
-                    logger.error(`Event Listener: ${error}`);
-                    throw(error);
+): Promise<any> => {
+    const listener: BlockListener = async (event: BlockEvent) => {
+        let bh_db: DBConnector;
+        bh_db = new LevelDBConnector(DB_NAME!, DB_OPEN_TIMEOUT);
+        await bh_db.open();
+        try {
+            const lastBlockNum = await getLastReadBlockNumber(bh_db, channelId);
+            // Log failed events for debugging purpose
+            const currBlockNum = event.blockNumber.toNumber();
+            let validTxIdSet = new Set();
+            event.getTransactionEvents().forEach((txEvent) => {
+                logger.info(`BlockEvent Listener: Block #${currBlockNum} event; LastReadBlock #${lastBlockNum}\n
+                    TxId: ${txEvent.transactionId},
+                    TxStatus: ${txEvent.status},
+                    isTxValid: ${txEvent.isValid}\n`)
+                if (txEvent.isValid) {
+                    validTxIdSet.add(txEvent.transactionId);
+                    logger.debug(`BlockEvent Listener: Valid tx added: ${txEvent.transactionId}`);
                 }
-                await new Promise(f => setTimeout(f, 2000));    // Sleep 2 seconds
+            })
+            if (currBlockNum === lastBlockNum + 1) {
```

Comment:
Can `currBlockNum - lastBlockNum` not be greater than 1? In which case, shouldn't we parse all of those blocks?

Reply:
This is listener code, and we decided that if difference is greater than 1, then listener will discard and let monitor parse all those blocks, so that there is no duplication of events.

Reply:
Line 354 does that in Monitor function